src/Symfony/Component/HttpClient/Response/AmpResponse.php

```diff
@@ -88,30 +93,19 @@ public function __construct(AmpClientState $multi, Request $request, array $opti
             $onProgress((int) $info['size_download'], ((int) (1 + $info['download_content_length']) ?: 1) - 1, (array) $info);
         };
 
-        $this->id = $id = Loop::defer(static function () use ($request, $multi, &$id, &$info, &$headers, $canceller, &$options, $onProgress, &$handle, $logger) {
-            return self::generateResponse($request, $multi, $id, $info, $headers, $canceller, $options, $onProgress, $handle, $logger);
-        });
+        $throttle = new Success();
 
-        $info['pause_handler'] = static function (float $duration) use ($id, &$delay) {
-            if (null !== $delay) {
-                Loop::cancel($delay);
-                $delay = null;
-            }
+        $this->id = $id = Loop::defer(static function () use ($request, $multi, &$id, &$info, &$headers, $canceller, &$options, $onProgress, &$handle, $logger, &$throttle) {
+            yield $throttle;
 
-            if (0 < $duration) {
-                $duration += microtime(true);
-                Loop::disable($id);
-                $delay = Loop::defer(static function () use ($duration, $id, &$delay) {
-                    if (0 < $duration -= microtime(true)) {
-                        $delay = Loop::delay(ceil(1000 * $duration), static function () use ($id) { Loop::enable($id); });
-                    } else {
-                        $delay = null;
-                        Loop::enable($id);
-                    }
-                });
-            } else {
-                Loop::enable($id);
-            }
+            return new Coroutine(self::generateResponse($request, $multi, $id, $info, $headers, $canceller, $options, $onProgress, $handle, $logger, $throttle));
+        });
+
+        $info['pause_handler'] = static function (float $duration) use (&$throttle) {
+            $throttle = call(static function () use (&$throttle, $duration) {
+                yield $throttle;
+                yield delay(\ceil(1000 * $duration));
```

Comment:
when `$duration <= 0`, throttling should be disabled

Reply:
Should it completely disable throttling, so if the remaining pause is 0.5 seconds, should it unpause instantly?

Reply:
It should unpause without delays

Reply:
Done. :+1: